<a href="https://colab.research.google.com/github/maya-g-y/Final-Project-ML-Module--Spotify/blob/main/Spotify1_Data_Prep.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##Spotify Project
####Research objectives - Which musical features most strongly differentiate between genres

# Data Preperation

In [ ]:
# Import required library
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings("ignore")
%matplotlib inline

### 1. Uploading the data





In [ ]:
# Load the dataset from Kaggle (only one source is used, so no table merging is required)
import kagglehub

# Download latest version
path = kagglehub.dataset_download("joebeachcapital/30000-spotify-songs")

print("Path to dataset files:", path)

100%|██████████| 3.01M/3.01M [00:00<00:00, 73.5MB/s]

Extracting files...
Path to dataset files: /root/.cache/kagglehub/datasets/joebeachcapital/30000-spotify-songs/versions/2


In [ ]:
import os

# Display dataset information
print("Dataset path:", path)
print("Files in dataset:")
print(os.listdir(path))


Dataset path: /root/.cache/kagglehub/datasets/joebeachcapital/30000-spotify-songs/versions/2
Files in dataset:
['readme.md', 'spotify_songs.csv']


In [ ]:
#Look at the Data
df = pd.read_csv(f"{path}/spotify_songs.csv")
df.head()


track_id                                         track_name  \
0  6f807x0ima9a1j3VPbc7VN  I Don't Care (with Justin Bieber) - Loud Luxur...   
1  0r7CVbZTWZgbTCYdfa2P31                    Memories - Dillon Francis Remix   
2  1z1Hg7Vb0AhHDiEmnDE79l                    All the Time - Don Diablo Remix   
3  75FpbthrwQmzHlBJLuGdC7                  Call You Mine - Keanu Silva Remix   
4  1e8PAfcKUYoKkxPhrHqw4x            Someone You Loved - Future Humans Remix   

       track_artist  track_popularity          track_album_id  \
0        Ed Sheeran                66  2oCs0DGTsRO98Gh5ZSl2Cx   
1          Maroon 5                67  63rPSO264uRjW1X5E6cWv6   
2      Zara Larsson                70  1HoSmj2eLcsrR0vE9gThr4   
3  The Chainsmokers                60  1nqYsOef1yKKuGOVchbsk6   
4     Lewis Capaldi                69  7m7vv9wlQ4i0LFuJiE2zsQ   

                                    track_album_name track_album_release_date  \
0  I Don't Care (with Justin Bieber) [Loud Luxury...               2019-06-14   
1                    Memories (Dillon Francis Remix)               2019-12-13   
2                    All the Time (Don Diablo Remix)               2019-07-05   
3                        Call You Mine - The Remixes               2019-07-19   
4            Someone You Loved (Future Humans Remix)               2019-03-05   

  playlist_name             playlist_id playlist_genre  ... key  loudness  \
0     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop  ...   6    -2.634   
1     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop  ...  11    -4.969   
2     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop  ...   1    -3.432   
3     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop  ...   7    -3.778   
4     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop  ...   1    -4.672   

   mode  speechiness  acousticness  instrumentalness  liveness  valence  \
0     1       0.0583        0.1020          0.000000    0.0653    0.518   
1     1       0.0373        0.0724          0.004210    0.3570    0.693   
2     0       0.0742        0.0794          0.000023    0.1100    0.613   
3     1       0.1020        0.0287          0.000009    0.2040    0.277   
4     1       0.0359        0.0803          0.000000    0.0833    0.725   

     tempo  duration_ms  
0  122.036       194754  
1   99.972       162600  
2  124.008       176616  
3  121.956       169093  
4  123.976       189052  

[5 rows x 23 columns]

In [ ]:
print("Shape of Dataset :",df.shape)

Shape of Dataset : (32833, 23)


In [ ]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32833 entries, 0 to 32832
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   track_id                  32833 non-null  object 
 1   track_name                32828 non-null  object 
 2   track_artist              32828 non-null  object 
 3   track_popularity          32833 non-null  int64  
 4   track_album_id            32833 non-null  object 
 5   track_album_name          32828 non-null  object 
 6   track_album_release_date  32833 non-null  object 
 7   playlist_name             32833 non-null  object 
 8   playlist_id               32833 non-null  object 
 9   playlist_genre            32833 non-null  object 
 10  playlist_subgenre         32833 non-null  object 
 11  danceability              32833 non-null  float64
 12  energy                    32833 non-null  float64
 13  key                       32833 non-null  int64  
 14  loudne

In [ ]:
# Create a copy of the original dataframe to preserve the raw data
df_prep= df.copy()

In [ ]:
# Select only columns with dtype = object
object_cols = df_prep.select_dtypes(include=['object'])

# Print the names of the object features
print("Object type features:")
print(object_cols.columns.tolist())




Object type features:
['track_id', 'track_name', 'track_artist', 'track_album_id', 'track_album_name', 'track_album_release_date', 'playlist_name', 'playlist_id', 'playlist_genre', 'playlist_subgenre']


In [ ]:
# Show the first rows of these features
object_cols.head()

track_id                                         track_name  \
0  6f807x0ima9a1j3VPbc7VN  I Don't Care (with Justin Bieber) - Loud Luxur...   
1  0r7CVbZTWZgbTCYdfa2P31                    Memories - Dillon Francis Remix   
2  1z1Hg7Vb0AhHDiEmnDE79l                    All the Time - Don Diablo Remix   
3  75FpbthrwQmzHlBJLuGdC7                  Call You Mine - Keanu Silva Remix   
4  1e8PAfcKUYoKkxPhrHqw4x            Someone You Loved - Future Humans Remix   

       track_artist          track_album_id  \
0        Ed Sheeran  2oCs0DGTsRO98Gh5ZSl2Cx   
1          Maroon 5  63rPSO264uRjW1X5E6cWv6   
2      Zara Larsson  1HoSmj2eLcsrR0vE9gThr4   
3  The Chainsmokers  1nqYsOef1yKKuGOVchbsk6   
4     Lewis Capaldi  7m7vv9wlQ4i0LFuJiE2zsQ   

                                    track_album_name track_album_release_date  \
0  I Don't Care (with Justin Bieber) [Loud Luxury...               2019-06-14   
1                    Memories (Dillon Francis Remix)               2019-12-13   
2                    All the Time (Don Diablo Remix)               2019-07-05   
3                        Call You Mine - The Remixes               2019-07-19   
4            Someone You Loved (Future Humans Remix)               2019-03-05   

  playlist_name             playlist_id playlist_genre playlist_subgenre  
0     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop  
1     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop  
2     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop  
3     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop  
4     Pop Remix  37i9dQZF1DXcZDD7cfEKhW            pop         dance pop

### 2. Convert an objects into viable data

In [ ]:
# 1. Convert all object columns to string EXCEPT track_album_release_date
obj_cols = df_prep.select_dtypes(include=['object']).columns.drop("track_album_release_date")
df_prep[obj_cols] = df_prep[obj_cols].astype("string")

print("Converted to string:")
print(obj_cols.tolist())

# 2. Convert track_album_release_date to datetime
df_prep["track_album_release_date"] = pd.to_datetime(
    df_prep["track_album_release_date"], errors="coerce"
)

# 3. Extract year and month as integers
df_prep["release_year"] = df_prep["track_album_release_date"].dt.year.astype("Int64")
df_prep["release_month"] = df_prep["track_album_release_date"].dt.month.astype("Int64")

# Quick check
print(df_prep[["track_album_release_date", "release_year", "release_month"]].head())




Converted to string:
['track_id', 'track_name', 'track_artist', 'track_album_id', 'track_album_name', 'playlist_name', 'playlist_id', 'playlist_genre', 'playlist_subgenre']
  track_album_release_date  release_year  release_month
0               2019-06-14          2019              6
1               2019-12-13          2019             12
2               2019-07-05          2019              7
3               2019-07-19          2019              7
4               2019-03-05          2019              3


In [ ]:
df_prep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32833 entries, 0 to 32832
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   track_id                  32833 non-null  string        
 1   track_name                32828 non-null  string        
 2   track_artist              32828 non-null  string        
 3   track_popularity          32833 non-null  int64         
 4   track_album_id            32833 non-null  string        
 5   track_album_name          32828 non-null  string        
 6   track_album_release_date  30947 non-null  datetime64[ns]
 7   playlist_name             32833 non-null  string        
 8   playlist_id               32833 non-null  string        
 9   playlist_genre            32833 non-null  string        
 10  playlist_subgenre         32833 non-null  string        
 11  danceability              32833 non-null  float64       
 12  energy            

### 3. Clean text

In [ ]:
import re
import string

# Define the columns to be cleaned
cols = [
    "playlist_name",
    "track_name",
    "track_album_name",
    "track_artist",
    "playlist_genre",
    "playlist_subgenre",
]

# ensure pandas StringDtype first
df_prep[cols] = df_prep[cols].astype("string")

# regex class for punctuation
punct_class = "[" + re.escape(string.punctuation) + "]"

for c in cols:
    s = df_prep[c].astype("string")
    # remove punctuation
    s = s.str.replace(punct_class, "", regex=True)
    # remove emojis / non-alphanumeric (keep letters, digits, whitespace)
    s = s.str.replace(r"[^A-Za-z0-9\s]", "", regex=True)
    # lowercase, trim, collapse multiple spaces
    s = s.str.lower().str.strip().str.replace(r"\s+", " ", regex=True)
    df_prep[c] = s.astype("string")


# verify dtypes
print(df_prep[cols].dtypes)




playlist_name        string[python]
track_name           string[python]
track_album_name     string[python]
track_artist         string[python]
playlist_genre       string[python]
playlist_subgenre    string[python]
dtype: object


In [ ]:
df_prep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32833 entries, 0 to 32832
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   track_id                  32833 non-null  string        
 1   track_name                32828 non-null  string        
 2   track_artist              32828 non-null  string        
 3   track_popularity          32833 non-null  int64         
 4   track_album_id            32833 non-null  string        
 5   track_album_name          32828 non-null  string        
 6   track_album_release_date  30947 non-null  datetime64[ns]
 7   playlist_name             32833 non-null  string        
 8   playlist_id               32833 non-null  string        
 9   playlist_genre            32833 non-null  string        
 10  playlist_subgenre         32833 non-null  string        
 11  danceability              32833 non-null  float64       
 12  energy            

### 4. Exploring features

In [ ]:
# Define column groups
id_cols = [c for c in df_prep.columns if "id" in c.lower()]
categorical_cols = [
    "playlist_name",
    "track_name",
    "track_album_name",
    "track_artist",
    "playlist_genre",
    "playlist_subgenre"
]
numeric_cols = df_prep.select_dtypes(include=["number"]).columns.tolist()
date_cols = ["track_album_release_date"]

# Explore relevant groups
text_cols = [c for c in categorical_cols if c not in id_cols]

for col in text_cols:
    print(f"\nColumn: {col}")
    print("Unique values:", df_prep[col].nunique(dropna=False))

    vc = df_prep[col].value_counts(dropna=False)

    print("\nTop 5 categories:")
    print(vc.head(5))

    print("\nBottom 5 categories:")
    print(vc.tail(5))



Column: playlist_name
Unique values: 423

Top 5 categories:
playlist_name
indie poptimism       579
permanent wave        514
urban contemporary    457
post teen pop         317
electropop            308
Name: count, dtype: Int64

Bottom 5 categories:
playlist_name
intro to permanent wave                        8
big white roomjessiej                          7
csr 1039 gta sa                                7
top 50 global 2020 updated weekly worldwide    6
alpas music festival                           3
Name: count, dtype: Int64

Column: track_name
Unique values: 22802

Top 5 categories:
track_name
           168
poison      22
breathe     21
alive       20
forever     20
Name: count, dtype: Int64

Bottom 5 categories:
track_name
its you radio edit                      1
reality feat janieck devy radio edit    1
dancing in the headlights radio edit    1
cool girl                               1
smooth feat rob thomas                  1
Name: count, dtype: Int64

Column: track_album_

In [ ]:
# Explore categorical features: unique counts, top/bottom categories, and rows affected by custom thresholds

def explore_categories_many(df, cols=None, default_min_count=30, per_col_threshold=None, top_n=5, bottom_n=5):
    """
    Explore multiple categorical columns with custom thresholds per column.
    per_col_threshold: dict like {"playlist_name": 40, "track_artist": 2}
    """

    if cols is None:
        cols = [c for c in categorical_cols if c not in id_cols]


    if per_col_threshold is None:
        per_col_threshold = {}

    for col in cols:
        s = df[col]
        n_rows = len(s)
        vc = s.value_counts(dropna=False)
        n_unique = vc.size

        # choose threshold: specific for this col, else default
        th = per_col_threshold.get(col, default_min_count)

        # Identify categories with frequency below the threshold
        below_mask = vc < th
        n_cats_below = below_mask.sum()      # Number of categories below threshold
        rows_below = vc[below_mask].sum()    # Total rows affected if we remove them
        pct_rows_below = (rows_below / n_rows * 100) if n_rows else 0.0  # Percentage of affected rows

        # Print exploration summary for this column
        print(f"\n=== Explore: {col} ===")
        print(f"Unique categories: {n_unique}")
        print(f"Categories below {th}: {n_cats_below}")
        print(f"Rows affected if reduced now: {rows_below} ({pct_rows_below:.2f}%)")

        # Display the most and least frequent categories
        print("\nTop categories:")
        print(vc.head(top_n))
        print("\nBottom categories:")
        print(vc.tail(bottom_n))

# Define custom frequency thresholds per column (used in the exploration function)
thresholds = {
    "playlist_name": 50,
    "track_name": 2,
    "track_artist": 2,
    "track_album_name": 2,
}



In [ ]:
# Just inspect first
explore_categories_many(
    df_prep,
    default_min_count=30,
    per_col_threshold=thresholds
)



=== Explore: playlist_name ===
Unique categories: 423
Categories below 50: 112
Rows affected if reduced now: 3606 (10.98%)

Top categories:
playlist_name
indie poptimism       579
permanent wave        514
urban contemporary    457
post teen pop         317
electropop            308
Name: count, dtype: Int64

Bottom categories:
playlist_name
intro to permanent wave                        8
big white roomjessiej                          7
csr 1039 gta sa                                7
top 50 global 2020 updated weekly worldwide    6
alpas music festival                           3
Name: count, dtype: Int64

=== Explore: track_name ===
Unique categories: 22802
Categories below 2: 17686
Rows affected if reduced now: 17686 (53.87%)

Top categories:
track_name
           168
poison      22
breathe     21
alive       20
forever     20
Name: count, dtype: Int64

Bottom categories:
track_name
its you radio edit                      1
reality feat janieck devy radio edit    1
dancing in the 

| Column              | Unique Categories | Categories Below Threshold | % Rows Affected if Reduced | Top Category (count, %)             | Bottom Example (count, %)            | Decision                                |
|---------------------|-------------------|----------------------------|----------------------------|-------------------------------------|--------------------------------------|-----------------------------------------|
| **playlist_name**   | 423               | 112 (<50)                  | 10.98%                     | Indie Poptimism (579, 1.76%)        | Alpas Music Festival (3, 0.01%)      | Reduce rare playlists to 'other'        |
| **track_name**      | 22,802            | 17,686 (<2)                | 53.87%                     | — (168, 0.51%)                      | Smooth (1, 0.00%)                     | Too sparse, likely to drop after EDA    |
| **track_album_name**| 19,332            | 13,749 (<2)                | 41.88%                     | Greatest Hits (142, 0.43%)          | Blackened Eyes (1, 0.00%)             | Too sparse, likely to drop after EDA    |
| **track_artist**    | 10,652            | 6,101 (<2)                 | 18.58%                     | Martin Garrix (161, 0.49%)          | Marky Mark & The Funky Bunch (1, 0.00%) | Keep with reduction (rare → other)     |
| **playlist_genre**  | 6                 | 0 (<30)                    | 0.00%                      | EDM (6043, 18.4%)                   | Rock (4951, 15.1%)                    | Keep as-is (no reduction needed)        |
| **playlist_subgenre**| 24               | 0 (<30)                    | 0.00%                      | Progressive Electro House (1809, 5.5%) | Reggaeton (949, 2.9%)                 | Keep as-is (no reduction needed)        |


### 5. Narrowing Categories

In [ ]:
# Function to reduce rare categories in a chosen column by grouping them into 'other'

def reduce_categories(df, col, min_count=30, other_label="other"):
    """
    Reduce rare categories in a column to 'other'.

    Parameters:
        df (pd.DataFrame): The dataframe to modify.
        col (str): The column name to reduce.
        min_count (int): Minimum count to keep a category. Categories with fewer will be set to 'other'.
        other_label (str): The label for reduced categories.

    Returns:
        pd.DataFrame: The dataframe with reduced categories.
    """
    vc = df[col].value_counts(dropna=False)

    # keep categories with count >= min_count (inclusive)
    keep = set(vc[vc >= min_count].index)

    # count before reduction
    n_unique_before = df[col].nunique(dropna=False)

    # do not touch NaN; only replace values not in 'keep'
    mask_keep_or_nan = df[col].isin(keep) | df[col].isna()
    df[col] = df[col].where(mask_keep_or_nan, other_label)

    # count after reduction
    n_unique_after = df[col].nunique(dropna=False)
    n_other = (df[col] == other_label).sum()

    print(f"\nColumn: {col}")
    print(f"Unique before: {n_unique_before}")
    print(f"Unique after:  {n_unique_after}")
    print(f"Rows labeled as '{other_label}': {n_other}")

    return df



In [ ]:
df_test2 = df_prep.copy(deep=True)

In [ ]:
# Reduce on the copy#2
df_test2 = reduce_categories(df_test2, "playlist_name", min_count=50)
df_test2 = reduce_categories(df_test2, "track_artist", min_count=2)


Column: playlist_name
Unique before: 423
Unique after:  312
Rows labeled as 'other': 3606

Column: track_artist
Unique before: 10652
Unique after:  4552
Rows labeled as 'other': 6101


In [ ]:
# Apply to the real df
df_prep= df_test2

In [ ]:
df_prep.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32833 entries, 0 to 32832
Data columns (total 25 columns):
 #   Column                    Non-Null Count  Dtype         
---  ------                    --------------  -----         
 0   track_id                  32833 non-null  string        
 1   track_name                32828 non-null  string        
 2   track_artist              32828 non-null  string        
 3   track_popularity          32833 non-null  int64         
 4   track_album_id            32833 non-null  string        
 5   track_album_name          32828 non-null  string        
 6   track_album_release_date  30947 non-null  datetime64[ns]
 7   playlist_name             32833 non-null  string        
 8   playlist_id               32833 non-null  string        
 9   playlist_genre            32833 non-null  string        
 10  playlist_subgenre         32833 non-null  string        
 11  danceability              32833 non-null  float64       
 12  energy            

In [ ]:
numeric_cols

['track_popularity',
 'danceability',
 'energy',
 'key',
 'loudness',
 'mode',
 'speechiness',
 'acousticness',
 'instrumentalness',
 'liveness',
 'valence',
 'tempo',
 'duration_ms',
 'release_year',
 'release_month']

##Save to pickle

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

# Save a permanent copy in Google Drive
df_prep.to_pickle("/content/drive/MyDrive/pickle_files/spotify_flat_file.pkl")

print("Saved cleaned dataframe to Google Drive.")


Mounted at /content/drive
Saved cleaned dataframe to Google Drive.


# We got A flat file, Next to EDA